src/quorumnet/bt_serialize.h

```diff
@@ -0,0 +1,603 @@
+using bt_value = boost::make_recursive_variant<
+    std::string, int64_t,
+    std::list<boost::recursive_variant_>,
+    std::unordered_map<std::string, boost::recursive_variant_>
+>::type;
+
+/// Convenience type that holds a dict of generic values (though *any* compatible data type can be used).
+using bt_dict = std::unordered_map<std::string, bt_value>;
+
+/// Convenience type that holds a list of generic values (though *any* compatible data type can be used).
+using bt_list = std::list<bt_value>;
+
+
+namespace detail {
+
+// Fallback base case; we only get here if none of the partial specializations below work
+template <typename T, typename SFINAE = void>
+struct bt_serialize { static_assert(!std::is_same<T, T>::value, "Cannot serialize T: unsupported type for bt serialization"); };
+
+template <typename T, typename SFINAE = void>
+struct bt_deserialize { static_assert(!std::is_same<T, T>::value, "Cannot deserialize T: unsupported type for bt deserialization"); };
+
+/// Checks that we aren't at EOF (or other !good() status) and throws if we are.
+inline void bt_no_eof(std::istream &is) {
+    if (!is.good())
+        throw bt_deserialize_invalid(
+                std::string(is.eof() ? "Unexpected EOF" : "I/O error") + " while deserializing at location " +
+                std::to_string(is.tellg()));
+}
+
+/// Checks that the type code is the next character, and if so, consumes it.  If not, throws.
+template <typename Exception = bt_deserialize_invalid>
+void bt_expect(std::istream &is, char expect) {
+    char t;
+    is.get(t);
+    bt_no_eof(is);
+    if (t != expect) {
+        is.unget();
+        throw Exception("Expected '" + std::string(1, expect) + "' but found '" + std::string(1, t) + "' at input location " + std::to_string(is.tellg()));
+    }
+}
+
+extern template void bt_expect<bt_deserialize_invalid>(std::istream &is, char expect);
+extern template void bt_expect<bt_deserialize_invalid_type>(std::istream &is, char expect);
+
+union maybe_signed_int64_t { int64_t i64; uint64_t u64; };
+
+/// Deserializes a signed or unsigned 64.  Sets the second bool to true iff the value is int64_t
+/// because a negative value was read.  Throws an exception if the read value doesn't fit in a
+/// int64_t (if negative) or a uint64_t (if positive).
+std::pair<maybe_signed_int64_t, bool> bt_deserialize_integer(std::istream &is);
+
+/// Integer specializations
+template <typename T>
+struct bt_serialize<T, enable_if_t<std::is_integral<T>::value>> {
+    static_assert(sizeof(T) <= sizeof(uint64_t), "Serialization of integers larger than uint64_t is not supported");
+    void operator()(std::ostream &os, const T &val) {
+        // Cast 1-byte types to a larger type to avoid iostream interpreting them as single characters
+        using output_type = typename std::conditional<(sizeof(T) > 1), T,
+              typename std::conditional<std::is_signed<T>::value, int, unsigned>::type>::type;
```

Comment:
The reason for the first part of the conditional is to make the `static_cast` a no-op for anything except 1-byte integer types; the nested conditional is to make it signed/unsigned based on the starting type (and I don't see how that one could be avoided).

C++14 lets this be a little more readable though:
```C++
using output_type = std::conditional_t<(sizeof(T) > 1), T, std::conditional_t<std::is_signed<T>::value, int, unsigned>>;
```

Reply:
I wonder because I thought zero extending is practically free. No-op is better than zero extending though. But readability and well, templates. Well not the end of the world.